tensorflow/stream_executor/rocm/rocm_blas.cc

```diff
@@ -1877,37 +1880,70 @@ port::Status ReorganizeMemory(Stream* stream,
   size_t matrix_byte_size = batch_stride * sizeof(MAPPED_T);  
   uint64_t cur_stride_size = matrix_byte_size;
 
+  struct write_op
+  {
+    char* src_ptr;
+    char* dst_ptr;
+    uint64_t size;
+  };
+  std::vector<write_op> write_ops;
+
   for (int i = 1; i < batch_count; ++i) {
     if (reinterpret_cast<char*>(raw_ptrs[i]) == src_ptr + cur_stride_size) {
       cur_stride_size += matrix_byte_size;
     } else {
-      DeviceMemoryBase src_mem = DeviceMemoryBase(src_ptr, cur_stride_size);
-      DeviceMemoryBase target_mem = DeviceMemoryBase(dst_ptr, cur_stride_size);
+      write_ops.push_back(write_op{src_ptr,dst_ptr,cur_stride_size});
+      src_ptr = reinterpret_cast<char *>(raw_ptrs[i]);
+      dst_ptr = device_memory_ptr + i * matrix_byte_size;
+      cur_stride_size = matrix_byte_size;
+    }
+  }  
+  write_ops.push_back(write_op{src_ptr,dst_ptr,cur_stride_size});
+
+  // if the requested operation is reducible to a broadcast (copying the same
+  // buffer multiple times to N uniformly distributed destinations), it may be
+  // more efficient to launch a single broadcast kernel than to launch N
+  // hipMemcpy's.
+  //  
+  // The broadcast kernel is not guaranteed to be as optimized as hipMemcpy,
+  // therefore, only take the shortcut if we're expecting high launch overhead
+  // (TODO: benchmark this to pick the optimal parameters?)
+  if(write_ops.size()>=4
+      && write_ops[0].size < 10000000
+      && !(write_ops[0].size & 3)) {
```

Comment:
Nit: Need to use a variable with better naming. It is hard for another developer to understand those magic constants.